### PR TITLE
Rules

### PR DESCRIPTION
**Pull request by DiegoVinasco**
- add proof of concept OWL+SWRL rule (reasoning-based) test suite #23
  - add basic rules for testing temporal consistency over OWL-Time entites
  - add basic rules for testing citygml versioning consistency according to [Samuel 2020]
  - validate gratte ciel data as consistent
- update ontologies with universally quantified constraints #24 
- integrate OWL-Time into the gratte ciel data #25 
  - also propose a small extension to OWL-Time for comparing instants and intervals

### CODE REVIEW

File: rules/swrl_rules.py [hunk 1]
```diff
@@ -0,0 +1,160 @@
+import json
+from owlready2 import sync_reasoner_pellet, World, Ontology, Imp, Nothing
+import logging
+from rdflib import Graph, Namespace
+
+
```
Comment (johnsamuelwrites) [inline rules/swrl_rules.py:6]: Please add docstring to these functions. It is not clear why the first ontology in the list is an ontology network and the remaining are ontology locations.

File: rules/ntriplify.py [hunk 2]
```diff
@@ -0,0 +1,11 @@
+### A simple script for converting all Turtle RDF files in the current directory to the N-triples RDF format
+
+import os
```
Comment (johnsamuelwrites) [inline rules/ntriplify.py:3]: Is it possible to specify the list of files from the command line. You may use argparse for parsing the list of RDF files. You can also specify an option for a directory (containing the list of RDF files).

File: Transformations/utilities/reformat_graph.py [hunk 3]
```diff
@@ -22,17 +21,9 @@ class RdfConverter():
     def __init__(self):
         self.graph = rdflib.Graph()
 
-    def convert(self, input_file, input_format, output_file='', output_format='xml'):
```
Comment (johnsamuelwrites) [inline Transformations/utilities/reformat_graph.py:25]: Missing docstrings in this file.

File: Transformations/utilities/reformat_graph.py [hunk 4]
```diff
@@ -22,17 +21,9 @@ class RdfConverter():
...
+    def convert(self, input_file, input_format, output_file, output_format):
         self.graph.parse(input_file, format=input_format)
-        if output_file == '':
```
Comment (johnsamuelwrites) [inline Transformations/utilities/reformat_graph.py:27]: It's not clear what's the input format and the output format from the method name.

File: Transformations/utilities/AddTimeStamps.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
+import logging
```
Comment (johnsamuelwrites) [inline Transformations/utilities/AddTimeStamps.py:1]: Please precise in the beginning of the file the need for adding timestamps. Missing docstrings and comments for the functions.

File: Transformations/ShapeChange/ontologyPatcher.py [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+    
+    patcher = ontologyPatcher(args.input_file, args.input_format, args.log)
+    patcher.patchOnClass()
+    patcher.patchDatatypeEnumeration()
+    patcher.patchObjectProperties()
+    patcher.outputGraph(args.output_file, args.output_format)
+
+
+class ontologyPatcher():
+    """A tool for patching ontologies created using ShapeChange.
```
Comment (johnsamuelwrites) [inline Transformations/ShapeChange/ontologyPatcher.py:31]: Why do we require this patching?

File: Transformations/DifferenceGraph-to-CityGML3/DifferenceGraph2RDF.py [hunk 7]
```diff
@@ -88,59 +98,106 @@ def main():
...
     transaction_member_uri = URIRef( VERS.VersionTransition + '.transaction' )
     from_uri = URIRef( VERS.VersionTransition + '.from' )
     to_uri = URIRef( VERS.VersionTransition + '.to' )
-    output_graph.add( (versiontransition_uri, from_uri, source_version_uri) )
-    output_graph.add( (versiontransition_uri, to_uri, target_version_uri) )
+    output_graph.add( (versionTransition_uri, from_uri, source_version_uri) )
+    output_graph.add( (versionTransition_uri, to_uri, target_version_uri) )
+
+    # add version and versionTransition timestamps if provided
+    if args.existence_time_stamps is not None:
```
Comment (johnsamuelwrites) [inline Transformations/DifferenceGraph-to-CityGML3/DifferenceGraph2RDF.py:118]: Add detailed comments to explain the if conditions.

File: Transformations/DifferenceGraph-to-CityGML3/testDifference.sh [hunk 8]
```diff
@@ -1,37 +1,57 @@
 #!/bin/sh
 
+python DifferenceGraph2RDF.py \
```
Comment (johnsamuelwrites) [inline Transformations/DifferenceGraph-to-CityGML3/testDifference.sh:3]: Add comments

File: Transformations/utilities/testTimestamp.sh [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+    -d
+
+python AddTimeStamps.py \
+    ../test-data/RDF/historicalSuccession_CityGML_3.0_v2.ttl \
+    ./historicalSuccession_v2.ttl \
+    2001-01-01T00:00:00 2001-01-01T00:00:00 \
+    -d
+
+
+# python AddTimeStamps.py \
```
Comment (johnsamuelwrites) [inline Transformations/utilities/testTimestamp.sh:16]: Are the following commented lines important? If not, these could be improved.

File: Transformations/utilities/testTimestamp.sh [hunk 10]
```diff
@@ -0,0 +1,67 @@
+#!/bin/sh
+
```
Comment (johnsamuelwrites) [inline Transformations/utilities/testTimestamp.sh:2]: Add comments explaining the purpose.

**Review by johnsamuelwrites — COMMENTED** [review]
Great to see these changes towards the generation of graph data. However, the current version misses docstrings in the Python files and comments in the shell scripts. I feel that we could include some (small) examples in the different README files.

File: rules/swrl_rules.py (hunk 1 above)
Comment (DiegoVinasco) [reply rules/swrl_rules.py:6]: added docstrings, see [line 9](https://github.com/VCityTeam/UD-Graph/blob/aa846ff115fd7ce7eb5a3348a11464d4a3b35351/rules/swrl_rules.py#L9) about the ontology network. Ontology networks don't exist in OWLReady2 but you can import multiple ontologies into an existing ontology which (assuming alignments exist) is a network.

Alternatively, we can change this variable name to just ontology to be clearer

File: rules/ntriplify.py (hunk 2 above)
Comment (DiegoVinasco) [reply rules/ntriplify.py:1]: added [here](https://github.com/VCityTeam/UD-Graph/blob/aa846ff115fd7ce7eb5a3348a11464d4a3b35351/rules/ntriplify.py#L5)
created an [issue](https://github.com/VCityTeam/UD-Graph/issues/32) for directory option.

File: Transformations/utilities/reformat_graph.py (hunk 3 above)
Comment (DiegoVinasco) [reply Transformations/utilities/reformat_graph.py:25]: added docstrings [here](https://github.com/VCityTeam/UD-Graph/blob/aa846ff115fd7ce7eb5a3348a11464d4a3b35351/Transformations/utilities/reformat_graph.py#L26)

File: Transformations/utilities/reformat_graph.py (hunk 4 above)
Comment (DiegoVinasco) [reply Transformations/utilities/reformat_graph.py:27]: I tried to be more specific [here](https://github.com/VCityTeam/UD-Graph/blob/aa846ff115fd7ce7eb5a3348a11464d4a3b35351/Transformations/utilities/reformat_graph.py#L30)

File: Transformations/utilities/AddTimeStamps.py (hunk 5 above)
Comment (DiegoVinasco) [reply Transformations/utilities/AddTimeStamps.py:1]: added precision in the [argparse description](https://github.com/VCityTeam/UD-Graph/blob/aa846ff115fd7ce7eb5a3348a11464d4a3b35351/Transformations/utilities/AddTimeStamps.py#L8) and the [readme](https://github.com/VCityTeam/UD-Graph/blob/rules/Transformations/utilities/README.md#addtimestamps)

File: Transformations/ShapeChange/ontologyPatcher.py (hunk 6 above)
Comment (DiegoVinasco) [reply Transformations/ShapeChange/ontologyPatcher.py:31]: added precision in [readme](https://github.com/VCityTeam/UD-Graph/tree/rules/Transformations/ShapeChange#readme)
and patcher [functions](https://github.com/VCityTeam/UD-Graph/blob/aa846ff115fd7ce7eb5a3348a11464d4a3b35351/Transformations/ShapeChange/ontologyPatcher.py#L50)

File: Transformations/DifferenceGraph-to-CityGML3/DifferenceGraph2RDF.py (hunk 7 above)
Comment (DiegoVinasco) [reply Transformations/DifferenceGraph-to-CityGML3/DifferenceGraph2RDF.py:124]: added explination in the [argparse documentation](https://github.com/VCityTeam/UD-Graph/blob/aa846ff115fd7ce7eb5a3348a11464d4a3b35351/Transformations/DifferenceGraph-to-CityGML3/DifferenceGraph2RDF.py#L19) for each type of timestamp

File: Transformations/ShapeChange/patchOntologies.sh [hunk 11]
```diff
@@ -0,0 +1,43 @@
+#!/bin/sh
+
+### CityGML 2.0
```
[the remark this answers is not in the document]
Comment (DiegoVinasco) [reply Transformations/ShapeChange/patchOntologies.sh]: moved script to [readme](https://github.com/VCityTeam/UD-Graph/tree/rules/Transformations/ShapeChange#to-run-the-ontology-patcher)

File: Transformations/utilities/testTimestamp.sh (hunk 9 above)
Comment (DiegoVinasco) [reply Transformations/utilities/testTimestamp.sh]: moved script to [readme](https://github.com/VCityTeam/UD-Graph/blob/rules/Transformations/utilities/README.md#addtimestamps)

File: Transformations/utilities/testTimestamp.sh (hunk 10 above)
Comment (DiegoVinasco) [reply Transformations/utilities/testTimestamp.sh]: updated documentation in [readme](https://github.com/VCityTeam/UD-Graph/blob/rules/Transformations/utilities/README.md#addtimestamps)

File: Transformations/DifferenceGraph-to-CityGML3/readme.md [hunk 12]
```diff
@@ -1,28 +1,48 @@
 # Difference Graph to CityGML v3.0 Versioning
+Data treatment scripts for transforming difference or change detection graphs from GraphML (JSON) to CityGML 3.0 Versioning Ontology individuals (in RDF formats)
 
-This script can take a difference or change detection graph as JSON and convert it into CityGML 3.0 Versioning Ontology individuals
+
+## [DifferenceGraph2RDF.py](DifferenceGraph2RDF.py) documentation
 
 The difference graph is created by the stage 4 output of the [cityGMLto3DTiles](https://github.com/VCityTeam/cityGMLto3DTiles/tree/master/PythonCallingDocker#manual-step-by-step-run-of-the-temporal-tiler) temporal data pipeline. These JSON graphs contain the geometric and geographical changes between objects of two or more CityGML 2.0 vintages of a city.
```
[the remark this answers is not in the document]
Comment (DiegoVinasco) [reply Transformations/DifferenceGraph-to-CityGML3/readme.md:7]: added examples to [readme](https://github.com/VCityTeam/UD-Graph/tree/rules/Transformations/DifferenceGraph-to-CityGML3#differencegraph2rdfpy-documentation)

**Comment by DiegoVinasco** [comment]
@johnsamuelwrites, I've removed the saxon jar file from the repository. Instead some documentation for how to download and install the files are provided in the [readme.md](https://github.com/VCityTeam/UD-Graph/pull/31/commits/88401c0d2eb0d1174e4c59209bdca1f07b0a34c7#diff-177cbf2693f2faecf51742154f22f1d9c9abebb21946e74e6ecc259c16e61d73)

**Pull request merged**